lib/onc_certification_g10_test_kit/smart_invalid_pkce_group.rb

```diff
@@ -0,0 +1,305 @@
+module ONCCertificationG10TestKit
+  class InvalidSMARTTokenRequestTest < Inferno::Test
+    title 'OAuth token exchange fails when supplied invalid code_verifier'
+    description %(
+      If the request failed verification or is invalid, the authorization
+      server returns an error response.
+    )
+    uses_request :redirect
+    id :invalid_pkce_request
+
+    input :code, :use_pkce, :pkce_code_verifier, :client_id, :client_secret, :smart_token_url
+
+    def modify_oauth_params(oauth_params)
+      oauth_params
+    end
+
+    run do
+      skip_if request.query_parameters['error'].present?, 'Error during authorization request'
+
+      oauth2_params = {
+        grant_type: 'authorization_code',
+        code: code,
+        redirect_uri: config.options[:redirect_uri]
+      }
+
+      oauth2_headers = { 'Content-Type' => 'application/x-www-form-urlencoded' }
+
+      if client_secret.present?
+        client_credentials = "#{client_id}:#{client_secret}"
+        oauth2_headers['Authorization'] = "Basic #{Base64.strict_encode64(client_credentials)}"
+      else
+        oauth2_params[:client_id] = client_id
+      end
+
+      modify_oauth_params(oauth2_params)
+
+      post(smart_token_url, body: oauth2_params, name: :token, headers: oauth2_headers)
+
+      assert_response_status([400, 401])
+    end
+  end
+
+  class SMARTInvalidPKCEGroup < Inferno::TestGroup
+    title 'SMART App Launch Error: Invalid PKCE Code Verifier'
+    short_title 'SMART Invalid PKCE Code Verifier'
+    input_instructions %(
+      Register Inferno as a standalone application using the following information:
+
+      * Redirect URI: `#{SMARTAppLaunch::AppRedirectTest.config.options[:redirect_uri]}`
+    )
+    description %(
+      # Background
+
+      The #{title} Group verifies that a SMART Launch Sequence, specifically the
+      [Standalone
+      Launch](http://hl7.org/fhir/smart-app-launch/STU2/app-launch.html#launch-app-standalone-launch)
+      Sequence, does not work in the case where the client sends an invalid PKCE
+      `code_verifier`. This group performs four launches without a valid
+      `code_verifier` and verifies that these do not result in a successful
+      launch.
+
+      This test is not included as part of a regular SMART Launch Sequence
+      because some servers may not accept an authorization code after it has
+      been used unsuccessfully in this manner.
```

Comment:
```suggestion
      The #{title} Group verifies that a SMART Launch Sequence, specifically the
      [Standalone
      Launch](http://hl7.org/fhir/smart-app-launch/STU2/app-launch.html#launch-app-standalone-launch)
      Sequence, verifies that servers properly support PKCE.  It does this by ensuring the launch fails
      in the case where the client sends an invalid PKCE `code_verifier`.
      
      This group performs four launches with various forms of an invalid `code_verifier` 
      (e.g. incorrect `code_verifier`, blank `code_identifier`) and verifies that these do
      not result in a successful launch.  Testers can expect to be prompted four times
      that a redirect will occur in this test.

      This test is not included as part of the Single Patient App group
      because there is no way for a client to infer that PKCE is supported on the server
      properly without performing extra launches.  Attempting to verify this within the
      same launch cannot be done because some servers may not accept an authorization code
      after it has been used unsuccessfully in this manner.
```